Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift

```diff
@@ -165,15 +208,295 @@ struct IDEDiagnosticConsoleFormatter: DiagnosticConsoleFormatter {
     }
 }
 
-// FIXME: Improve the readability for diagnostics on the command line https://github.com/apple/swift-docc/issues/496
-struct DefaultDiagnosticConsoleFormatter: DiagnosticConsoleFormatter {
+// MARK: Default formatting
+
+final class DefaultDiagnosticConsoleFormatter: DiagnosticConsoleFormatter {
     var options: DiagnosticFormattingOptions
+    private let baseUrl: URL?
+    private let highlight: Bool
+    private var sourceLines: [URL: [String]] = [:]
+
+    /// The number of additional lines from the source file that should be displayed both before and after the diagnostic source line.
+    private static let contextSize = 2
     
-    func formattedDescription(for problem: Problem) -> String {
-        formattedDescription(for: problem.diagnostic)
+    init(
+        baseUrl: URL?,
+        highlight: Bool,
+        options: DiagnosticFormattingOptions
+    ) {
+        self.baseUrl = baseUrl
+        self.highlight = highlight
+        self.options = options
     }
     
+    func formattedDescription<Problems>(for problems: Problems) -> String where Problems: Sequence, Problems.Element == Problem {
+        let sortedProblems = problems.sorted { lhs, rhs in
+            guard let lhsSource = lhs.diagnostic.source,
+                  let rhsSource = rhs.diagnostic.source
+            else { return lhs.diagnostic.source  == nil }
+            
+            guard let lhsRange = lhs.diagnostic.range,
+                  let rhsRange = rhs.diagnostic.range
+            else { return lhsSource.path < rhsSource.path }
+            
+            if lhsSource.path == rhsSource.path {
+                return lhsRange.lowerBound < rhsRange.lowerBound
+            } else {
+                return lhsSource.path < rhsSource.path
+            }
+        }
+
+        return sortedProblems.map { formattedDescription(for: $0) }.joined(separator: "\n\n")
+    }
+
+    func formattedDescription(for problem: Problem) -> String {
+        defer { performCleanup() }
```

Comment:
It would be better to clean the cached source lines after all diagnostics have been written. Right now I don't think the cached values will be read since the cache is emptied between each diagnostic. 

I don't think formatter protocol has an API for this but you can probably add something similar to `finalize()` that the console writer could call in its `finalize` implementation.

Reply:
That's a great point, my bad! I just pushed a fix for this, thank you!